text/0000-guaranteed-slice-repr.md

```diff
@@ -0,0 +1,205 @@
+- Feature Name: guaranteed_slice_repr
+- Start Date: 2025-02-18
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC guarantees the in-memory representation of slice and str references.
+Specifically, `&[T]` and `&mut [T]` are guaranteed to have the same layout as:
+
+```rust
+#[repr(C)]
+struct Slice<T> {
+    data: *const T,
+    len: usize,
+}
+```
+
+The layout of `&str` is the same as that of `&[u8]`, and the layout of
+`&mut str` is the same as that of `&mut [u8]`.
+
+# Motivation
+[motivation]: #motivation
+
+This RFC allows non-Rust (e.g. C or C++) code to read from or write to existing
+slices and to declare slice fields or locals.
+
+For example, guaranteeing the representation of slice references allows
+non-Rust code to read from the `data` or `len` fields of `string` in the type
+below without intermediate FFI calls into Rust:
+
+```rust
+#[repr(C)]
+struct HasString {
+    string: &'static str,
+}
+```
+
+Note: prior to this RFC, the type above is not even properly `repr(C)` since the
+size and alignment of slices were not guaranteed. However, the Rust compiler
+accepts the `repr(C)` declaration above without warning.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Slice references are represented with a pointer and length pair. Their in-memory
+layout is the same as a `#[repr(C)]` struct like the following:
+
+```rust
+#[repr(C)]
+struct Slice<T> {
+    data: *const T,
+    len: usize,
```

Comment:
If we wished to extend Rust to allow references to unsized types that encompass unsized types... for example, `&[[T]]`, or more interestingly, `&[dyn Trait]`... then one of the likely choices we would make is to have "triple-wide" (or larger!) references, instead of this "double-wide" one. Thus we would only apply the corresponding layout this RFC describes (or any other such specified layout) in the case of `where T: Sized`.

I feel this "multiple metadata" possibility should be considered and either _explicitly reserved_ as a future possibility, as accepting this RFC could be interpreted as foreclosing it, or dismissed.

Reply:
I considered this, but IMO it seems difficult to imagine how `&[[T]]` would work-- even with triple-wide references, the items in the collection could be heterogeneous.

I suppose we could make it work if all of the elements were the same type, e.g. coercing `&[String]` into `&[dyn Debug]`.

Personally, I'd be happy to restrict this RFC to specify that this only applies to `&[T] where T: Sized`. Would that resolve your concern?

Reply:
we could make the RFC compatible with `&[[U]]` by simply relaxing the bound to `T: ?Sized`:

```rust
&[[U]]
->
{ data: *const [U], len: usize }
->
{ data: { data: *const U, len: usize }, len: usize }
```

so I don't see how this definition of `Slice<T>` forecloses the multi-meta possibility.

Reply:
> I considered this, but IMO it seems difficult to imagine how `&[[T]]` would work-- even with triple-wide references, the items in the collection could be heterogeneous.

You could either require homogeneity like you suggest, or you could have the metadata be its own pointer to a separate region of memory, sharing the lifetime of the data pointer. (I wrote a very experimental crate that does something like the latter: https://github.com/Jules-Bertholet/unsized-vec)

But since current Rust doesn’t support this, and nothing in this RFC would interfere with someday adopting either option, it’s not a concern IMO.

Reply:
> we could make the RFC compatible with `&[[U]]` by simply relaxing the bound to `T: ?Sized`

another valid extension is to just insert more fields for metadata between the pointer and length or change the length to be a struct-- these match how the pointer metadata APIs work more closely.

Reply:
> changing the metadata (length) from `usize` to a struct would be [rust-lang/libs-team#246](https://github.com/rust-lang/libs-team/issues/246)

I don't really think so, that's just wrapping the metadata type in a struct whenever you're not using it as part of a pointer type. This is explicitly similar to `mem::Discriminant` which wraps an enum's discriminant in a struct, even though the discriminant might be a well-known type, e.g. `u8` for a `repr(u8)` fieldless enum.

Reply:
I think both homogeneous `&[dyn Trait]` and rectangular `&[[T]]` sound useful, especially because the retangular case benefits from more optimizations.

We could consider unsizing options for more general types too like `struct Foo<T>(T,T)`, so then a homogeneous `Foo<dyn Trait>`.

Reply:
I think I lost the thread a bit here in terms of how this should affect the layout. AFAICT, there's no reasonable layout we can specify at this point for `&[T]` where `T` is not `Sized`. I'll change the RFC to specify that this layout guarantee is only applicable where `T: Sized`.

Reply:
> Personally, I'd be happy to restrict this RFC to specify that this only applies to &[T] where T: Sized. Would that resolve your concern?

Oh, yes, entirely.